# Exercise M1.03

The goal of this exercise is to compare the performance of our classifier in the previous notebook (roughly 81\% accuracy with `LogisticRegression`) to some simple baseline classifier. The simplest baseline classifier isone that always predicts the same class, irrespective of the input data.

* What would be the score of a model that always predicts ` >50K`?
* What would be the score of a model that always predicts ` <50k`?
* Is 81\% or 82\% accuracy a good score for this problem?

Use a `DummyClassifier` and do a train-test split to evaluate its accuracy on the test set. 

In [2]:
# Imports

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.dummy import DummyClassifier

In [3]:
df = pd.read_csv('data/adult-census.csv')
df.head()

age   workclass  fnlwgt      education  education-num       marital-status  \
0   25     Private  226802           11th              7        Never-married   
1   38     Private   89814        HS-grad              9   Married-civ-spouse   
2   28   Local-gov  336951     Assoc-acdm             12   Married-civ-spouse   
3   44     Private  160323   Some-college             10   Married-civ-spouse   
4   18           ?  103497   Some-college             10        Never-married   

           occupation relationship    race      sex  capital-gain  \
0   Machine-op-inspct    Own-child   Black     Male             0   
1     Farming-fishing      Husband   White     Male             0   
2     Protective-serv      Husband   White     Male             0   
3   Machine-op-inspct      Husband   Black     Male          7688   
4                   ?    Own-child   White   Female             0   

   capital-loss  hours-per-week  native-country   class  
0             0              40   United-States   <=50K  
1             0              50   United-States   <=50K  
2             0              40   United-States    >50K  
3             0              40   United-States    >50K  
4             0              30   United-States   <=50K

We first split our dataset to have the target separated from the data used to train our predictive model.

In [4]:
# remove fnlwgt and education
df = df.drop(columns=['fnlwgt', 'education'])

# select target
target = df['class']

# select data
data = df.drop(columns='class')

In [6]:
# since we will use only numerical data
data_numeric = data.select_dtypes(include='number')

In [7]:
# split the data
data_train, data_test, target_train, target_test = train_test_split(data_numeric, target, 
                                                                    test_size=.25, random_state=23)

Train the logistic regression model

In [12]:
model = LogisticRegression()
model.fit(data_train, target_train)
# score of the logistic regression model
model.score(data_test, target_test)

0.821390549504545

Use the `DummyClassifier` such that the resulting classifier will always predict the class ` >50K`.  
Repeat the experiment by always predicting the class ` <=50K`.

In [15]:
for constant in [' >50K', ' <=50K' ]:
    dummy_cl = DummyClassifier(strategy='constant', constant=constant)
    dummy_cl.fit(data_train, target_train)
    print(f"Score of a model that always predicts {constant}: ",
          f"{dummy_cl.score(data_test, target_test):.3f}")

Score of a model that always predicts  >50K:  0.233
Score of a model that always predicts  <=50K:  0.767


If we use the strategy `"most_frequent"` to predict the class that appears the most in the training target, which is in this example the same as ` <=50K`, we would be on average 77% of the time right, without even training a model.

However the LogisticRegression accuracy (roughly 82\%) is better than the `most_frequent` strategy.